pkg/distributor/ingest_limits.go

```diff
@@ -0,0 +1,154 @@
+
+	"github.com/grafana/dskit/ring"
+	ring_client "github.com/grafana/dskit/ring/client"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	limits_frontend_client "github.com/grafana/loki/v3/pkg/limits/frontend/client"
+	"github.com/grafana/loki/v3/pkg/logproto"
+)
+
+// ingestLimitsFrontendClient is used for tests.
+type ingestLimitsFrontendClient interface {
+	ExceedsLimits(context.Context, *logproto.ExceedsLimitsRequest) (*logproto.ExceedsLimitsResponse, error)
+}
+
+// ingestLimitsFrontendRingClient uses the ring to query ingest-limits frontends.
+type ingestLimitsFrontendRingClient struct {
+	ring ring.ReadRing
+	pool *ring_client.Pool
+}
+
+func newIngestLimitsFrontendRingClient(ring ring.ReadRing, pool *ring_client.Pool) *ingestLimitsFrontendRingClient {
+	return &ingestLimitsFrontendRingClient{
+		ring: ring,
+		pool: pool,
+	}
+}
+
+// Implements the ingestLimitsFrontendClient interface.
+func (c *ingestLimitsFrontendRingClient) ExceedsLimits(ctx context.Context, req *logproto.ExceedsLimitsRequest) (*logproto.ExceedsLimitsResponse, error) {
+	// We use an FNV-1 of all stream hashes in the request to load balance requests
+	// to limits-frontends instances.
+	h := fnv.New32()
+	for _, stream := range req.Streams {
+		// Add the stream hash to FNV-1.
+		buf := make([]byte, binary.MaxVarintLen64)
+		binary.PutUvarint(buf, stream.StreamHash)
+		_, _ = h.Write(buf)
+	}
+	// Get the limits-frontend instances from the ring.
+	var descs [5]ring.InstanceDesc
+	rs, err := c.ring.Get(h.Sum32(), limits_frontend_client.LimitsRead, descs[0:], nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get limits-frontend instances from ring: %w", err)
+	}
+	var lastErr error
+	// Send the request to the limits-frontend to see if it exceeds the tenant
+	// limits. If the RPC fails, failover to the next instance in the ring.
+	for _, instance := range rs.Instances {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+		c, err := c.pool.GetClientFor(instance.Addr)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		client := c.(logproto.IngestLimitsFrontendClient)
+		resp, err := client.ExceedsLimits(ctx, req)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		return resp, nil
+	}
+	return nil, lastErr
+}
+
+// exceedsIngestLimitsResult contains the reasons a stream exceeds per-tenant
+// ingest limits.
+type exceedsIngestLimitsResult struct {
+	hash    uint64
+	reasons []string
+}
+
+type ingestLimits struct {
+	client         ingestLimitsFrontendClient
+	limitsFailures prometheus.Counter
+}
+
+func newIngestLimits(client ingestLimitsFrontendClient, r prometheus.Registerer) *ingestLimits {
+	return &ingestLimits{
+		client: client,
+		limitsFailures: promauto.With(r).NewCounter(prometheus.CounterOpts{
+			Name: "loki_distributor_ingest_limits_failures_total",
+			Help: "The total number of failures checking per-tenant ingest limits.",
+		}),
```

Comment:
```suggestion
		limitsFailures: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
			Name: "loki_distributor_ingest_limits_failures_total",
			Help: "The total number of failures checking per-tenant ingest limits.",
		}, []string{"tenant"}),
```

Reply:
It's not the metric that is per-tenant, it's the service that is per-tenant (per-tenant ingest limits). I will fix the help message. R.E adding the `tenant` label, can I think on this? I still don't like how we will create a large number of (short) contiguous streams when a sudden failure happens. I wonder if adding the tenant would be better in a log message instead?